File: backend/calcul_btp/validators.py
import re

from django.core.exceptions import ValidationError


class ComplexPasswordValidator:
    """≥ 8 caractères, 1 majuscule, 1 minuscule, 1 chiffre."""

    def validate(self, password, user=None):
        if len(password) < 8:
            raise ValidationError("Au moins 8 caractères.", code="password_too_short")
        if not re.search(r"[A-Z]", password):
            raise ValidationError("Au moins une majuscule.", code="password_no_upper")
        if not re.search(r"[a-z]", password):
            raise ValidationError("Au moins une minuscule.", code="password_no_lower")
        if not re.search(r"[0-9]", password):
            raise ValidationError("Au moins un chiffre.", code="password_no_digit")

    def get_help_text(self):
        return (
            "Le mot de passe doit contenir au moins 8 caractères, "
            "une majuscule, une minuscule et un chiffre."
        )

File: lib/core/utils/validators.dart
abstract final class Validators {
  static String? required(String? value, [String label = 'Ce champ']) {
    if (value == null || value.trim().isEmpty) {
      return '$label est obligatoire.';
    }
    return null;
  }

  static String? email(String? value) {
    final base = required(value, 'L\'email');
    if (base != null) return base;
    final emailRegex = RegExp(r'^[^@\s]+@[^@\s]+\.[^@\s]+$');
    if (!emailRegex.hasMatch(value!.trim())) {
      return 'Adresse email invalide.';
    }
    return null;
  }

  /// ≥ 8 caractères, 1 majuscule, 1 minuscule, 1 chiffre.
  static String? password(String? value) {
    final base = required(value, 'Le mot de passe');
    if (base != null) return base;
    final v = value!;
    if (v.length < 8) return 'Au moins 8 caractères.';
    if (!RegExp(r'[A-Z]').hasMatch(v)) return 'Au moins une majuscule.';
    if (!RegExp(r'[a-z]').hasMatch(v)) return 'Au moins une minuscule.';
    if (!RegExp(r'[0-9]').hasMatch(v)) return 'Au moins un chiffre.';
    return null;
  }

  static String? positiveNumber(String? value, [String label = 'La valeur']) {
    final base = required(value, label);
    if (base != null) return base;
    final n = double.tryParse(value!.replaceAll(',', '.'));
    if (n == null) return 'Nombre invalide.';
    if (n <= 0) return '$label doit être > 0.';
    return null;
  }

  static double? parsePositive(String? value) {
    if (value == null || value.trim().isEmpty) return null;
    final n = double.tryParse(value.replaceAll(',', '.'));
    if (n == null || n <= 0) return null;
    return n;
  }
}
